R/admin.R

```diff
@@ -203,3 +203,99 @@ find_multiple_mutates <- function(path = "R/") {
     }
   }
 }
+
+
+#' normalize_files
+#'
+#' Normalize line endings for all package input data.
+#'
+#' @param root Folder root to scan, character
+#' @return Nothing - run for side effects only.
+#' @note Set \code{root} to "./extdata" in the git directory, not the package root, to make changes that 'stick'.
+#' @details Some GCAM input datafiles have bad line endings, and/or
+#' don't have a final newline. This utility script converts all files to have Unix line endings (\code{\\n}) and a final newline.
+#' @author BBL
+normalize_files <- function(root = system.file("extdata", package = "gcamdata")) {
+  assert_that(is.character(root))
+  message("Root: ", root)
+
+  # Get a list of all CSV input files
+  files <- list.files(root, pattern = "\\.csv$", full.names = TRUE, recursive = TRUE)
+
+  for(f in seq_along(files)) {
+    shortfn <- gsub(root, "", files[f])
+    size <- round(file.size(files[f]) / 1024 / 1024, 3)  # MB
+    message(f, "/", length(files), ": ", shortfn, ", ", size, " Mb ", appendLF = FALSE)
+
+    # Read file and then write it back out
+    message("\tReading...", appendLF = FALSE)
+    txt <- readLines(files[f], warn = FALSE)
+    uc_size <- format(utils::object.size(txt), units = "Mb")
+    message("OK. ", uc_size, " uncompressed")
+
+    message("\tWriting...", appendLF = FALSE)
+    writeLines(txt, files[f])
+    message("OK")
+  }
+}
+
+
+#' add_column_types_header_line
+#'
+#' One-off function to insert a `Column types` header line into input files.
+#'
+#' @return Nothing.
+add_column_types_header_line <- function() {
+
+  files <- list.files(path = "inst/extdata/", pattern = "csv$", full.names = TRUE, recursive = TRUE)
+
+  colchars <- c("character" = "c", "integer" = "i", "numeric" = "n", "double" = "d", "logical" = "I")
+
+  for(f in files) {
+    cat(f, "...")
+    header <- readLines(f, n = 20)
+    header <- header[grepl("^#", header)]
+    original_header_length <- length(header)
+
+    # Remove any previous Column types entry
```

Comment:
Should we have this function keep previous column type entries?  I'm thinking of the case that a user needs to go back and correct a CSV manually because `read.csv` didn't detect it correctly.  We wouldn't want to have to go back and redo the manually corrections every time we run this function.

On a maybe related note if I re-run this function on the latest commit I do get some files that show up modified, was that expected?:
```
diff --git a/inst/extdata/gcam-usa/UCS_Database.csv b/inst/extdata/gcam-usa/UCS_Database.csv
-# Column types: iccccccccnnicinicccnncccccc 
+# Column types: iccccccccnniciniccccccccccc 
diff --git a/inst/extdata/socioeconomics/SSP_database_v9.csv b/inst/extdata/socioeconomics/SSP_database_v9.csv
-# Column types: cccccccccccccccccnnnnnnnnnnnnnnnnnnncccccccccc 
+# Column types: cccccccccccccccnnnnnnnnnnnnnnnnnnnnncccccccccc 
diff --git a/inst/extdata/water/IBNET_municipal_water_cost_USDm3.csv b/inst/extdata/water/IBNET_municipal_water_cost_USDm3.csv
-# Column types: icnnn 
+# Column types: icncn 
```

Reply:
To the second question, yes, I had to hand-correct some files, so that's what you're seeing when you rerun. To the first...not sure. 😕